{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This notebook describes some steps used to detect CNV regions using SRBreak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####Load SRBreak into R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: CNVrd2\n",
       "Loading required package: VariantAnnotation\n",
       "Loading required package: BiocGenerics\n",
       "Loading required package: parallel\n",
       "\n",
       "Attaching package: ‘BiocGenerics’\n",
       "\n",
       "The following objects are masked from ‘package:parallel’:\n",
       "\n",
       "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
       "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
       "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
       "\n",
       "The following object is masked from ‘package:stats’:\n",
       "\n",
       "    xtabs\n",
       "\n",
       "The following objects are masked from ‘package:base’:\n",
       "\n",
       "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
       "    do.call, duplicated, eval, evalq, Filter, Find, get, intersect,\n",
       "    is.unsorted, lapply, Map, mapply, match, mget, order, paste, pmax,\n",
       "    pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rep.int,\n",
       "    rownames, sapply, setdiff, sort, table, tapply, union, unique,\n",
       "    unlist, unsplit\n",
       "\n",
       "Loading required package: GenomeInfoDb\n",
       "Loading required package: stats4\n",
       "Loading required package: S4Vectors\n",
       "Loading required package: IRanges\n",
       "Loading required package: GenomicRanges\n",
       "Loading required package: Rsamtools\n",
       "Loading required package: XVector\n",
       "Loading required package: Biostrings\n",
       "\n",
       "Attaching package: ‘VariantAnnotation’\n",
       "\n",
       "The following object is masked from ‘package:base’:\n",
       "\n",
       "    tabulate\n",
       "\n",
       "Loading required package: rjags\n",
       "Loading required package: coda\n",
       "Linked to JAGS 3.4.0\n",
       "Loaded modules: basemod,bugs\n",
       "Loading required package: ggplot2\n",
       "Loading required package: gridExtra\n",
       "Loading required package: grid\n",
       "Loading required package: mclust\n",
       "    __  ___________    __  _____________\n",
       "   /  |/  / ____/ /   / / / / ___/_  __/\n",
       "  / /|_/ / /   / /   / / / /\\__ \\ / /   \n",
       " / /  / / /___/ /___/ /_/ /___/ // /    \n",
       "/_/  /_/\\____/_____/\\____//____//_/    version 5.0.1\n",
       "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
       "\n",
       "Attaching package: ‘mclust’\n",
       "\n",
       "The following object is masked from ‘package:GenomicRanges’:\n",
       "\n",
       "    map\n",
       "\n",
       "The following object is masked from ‘package:IRanges’:\n",
       "\n",
       "    map\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "library(\"SRBreak\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Detect CNVs and their breakpoints "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###Example: single-end reads\n",
    "\n",
    "####Please download *BamSimulatedData* and extract it into the current directory\n",
    "\n",
    "####If you are using a dockerized version, you can go to the **root** directory and see the \"BamSimulatedData\" directory\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Loading required package: BSgenome\n",
       "Loading required package: rtracklayer\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L1000.101561000.101562000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L20000.101556000.101576000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L5000.101560000.101565000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Del.1.L84780.101545220.101630000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L10000.101558000.101568000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L20000.101556000.101576000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Dup.three.L50000.101555000.101605000.fa.9.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.10.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.11.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.12.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.13.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.14.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.15.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.1.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.2.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.3.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.4.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.5.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.6.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.7.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.8.x.bwa.sorted.bam\n",
       "Reading file: Normal.three.L84780.101545220.101630000.fa.9.x.bwa.sorted.bam\n",
       "=============================================\n",
       "120 bam files were read\n",
       "=============================================\n",
       "Correcting the GC content\n",
       "Correcting mappability bias\n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L1000.101561000.101562000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L20000.101556000.101576000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L5000.101560000.101565000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Del.1.L84780.101545220.101630000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L10000.101558000.101568000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L20000.101556000.101576000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Dup.three.L50000.101555000.101605000.fa.9.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.10.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.11.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.12.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.13.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.14.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.15.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.1.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.2.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.3.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.4.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.5.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.6.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.7.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.8.x.bwa.sorted.bam \n",
       "Analyzing: Normal.three.L84780.101545220.101630000.fa.9.x.bwa.sorted.bam \n",
       "Calculate segmentation scores for polymorphic regions\n",
       "   user  system elapsed \n",
       " 60.831   2.228  85.721 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "system.time(outputSRBreak <- SRBreak(readDepthWindow = 500,\n",
    "                                     chr = \"chr1\", ##Chromosome name\n",
    "                                                                                                            \n",
    "                                     st = 101100001, ##Start position\n",
    "                                     \n",
    "                                     en = 102100000, ##End position\n",
    "                                     \n",
    "                                     dirBamFile = \"./BamSimulatedData/\", ##Bam files' directory\n",
    "                                     \n",
    "                                     detectAllRegion = TRUE, ##Set this = TRUE in order to obtain all CNV regions\n",
    "                                     \n",
    "                                     rdQualityMapping = 0, ##Mapping quality (used in read-depth method)\n",
    "                        #mappabilityFile = \"mappabilitFile.chr1.start.101100001.end.102100000.window.500.txt\",\n",
    "                                     \n",
    "                                     testType = \"Count\", ##Test type\n",
    "                                     \n",
    "                                     correctGC = TRUE, ##Correct GC content\n",
    "                                     \n",
    "                                     upperCNThreshold = 0.25, ##Larger than this threshold is duplication\n",
    "                                     \n",
    "                                     lowerCNThreshold = -0.25, ##Smaller than this threshold is deletion\n",
    "                                     \n",
    "                                     countThreshold = 1, ##Number of duplications/deletions \n",
    "                                     \n",
    "                                     minLengthSV = 1000, ##Minimum length of a duplication/deletion event\n",
    "\n",
    "                                    #inputRawReadCountMatrix = readCountMatrix,\n",
    "                                     usingPairedEnds = FALSE, ##Not use paired-end information,\n",
    "                                                            \n",
    "                                     thresholdOfIntersectionBetweenRDandPEM = 0.9 #Not useful if only single-end reads used\n",
    "                                     \n",
    "                                    ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We can see all simulated breakpoints by extracting sample names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges of length 6\n",
       "        start       end width\n",
       "[1] 101545220 101630000 84781\n",
       "[2] 101555000 101605000 50001\n",
       "[3] 101556000 101576000 20001\n",
       "[4] 101558000 101568000 10001\n",
       "[5] 101560000 101565000  5001\n",
       "[6] 101561000 101562000  1001\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "sampleNames <- dir(\"BamSimulatedData\", \"bam$\")\n",
    "\n",
    "sampleOfSV <- sampleNames[grep(\"Dup|Del\", sampleNames)]\n",
    "sampleOfSV\n",
    "\n",
    "listSV <- unique(t(sapply(strsplit(sampleOfSV, \".\", fixed = TRUE), function(x) x[4:5])))\n",
    "    \n",
    "    \n",
    "simulatedSV  <- IRanges(as.integer(listSV[, 1]), as.integer(listSV[, 2]))\n",
    "    \n",
    "sort(simulatedSV)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "####Take a look at results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IRanges of length 7\n",
       "        start       end width\n",
       "[1] 101545221 101630001 84781\n",
       "[2] 101555000 101605001 50002\n",
       "[3] 101556001 101576005 20005\n",
       "[4] 101556005 101576001 19997\n",
       "[5] 101558001 101568001 10001\n",
       "[6] 101560001 101565000  5000\n",
       "[7] 101561001 101562001  1001\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%R\n",
    "rawoutputSV <- outputSRBreak$svResult\n",
    "\n",
    "outputSV <- IRanges(as.integer(rawoutputSV[, 2]), as.integer(rawoutputSV[, 3]))\n",
    "\n",
    "outputSV \n",
    "\n",
    "###Check whether the simulated breakpoints are in these results\n",
    "outputSV[(start(outputSV) > (101545220 - 5000)) & (end(outputSV) < (101630000 + 50000)) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
